File: 题库标准算法/2.两数相加/Java/Solution1.1.0.java
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode result = new ListNode();
        ListNode p = result;
        ListNode pl1 = l1;
        ListNode pl2 = l2;
        short carry = 0;

        while (pl1 != null || pl2 != null || carry != 0)
        {
            short sL1;
            short sL2;

            if (pl1 != null)
            {
                sL1 = (short) pl1.val;
            }
            else
            {
                sL1 = 0;
            }

            if (pl2 != null)
            {
                sL2 = (short) pl2.val;
            }
            else
            {
                sL2 = 0;
            }

            short value = (short) (sL1 + sL2 + carry);

            carry = (short) (value / 10);
            value %= 10;

            p.next = new ListNode(value);
            p = p.next;

            if (pl1 != null)
            {
                pl1 = pl1.next;
            }

            if (pl2 != null)
            {
                pl2 = pl2.next;
            }
        }

        return result.next;
    }
}

File: 题库标准算法/2.两数相加/Python/Solution1.1.0.py
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        result = ListNode()
        p = result
        pl1 = l1
        pl2 = l2
        carry = 0

        while pl1 or pl2 or carry:
            sl1 = 0
            sl2 = 0

            if pl1:
                sl1 = pl1.val

            if pl2:
                sl2 = pl2.val

            value = sl1 + sl2 + carry
            carry = value // 10
            value %= 10
            p.next = ListNode(value)
            p = p.next

            if pl1:
                pl1 = pl1.next

            if pl2:
                pl2 = pl2.next

        return result.next
